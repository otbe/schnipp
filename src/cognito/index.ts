import { CognitoUserPoolTriggerEvent } from 'aws-lambda';
import { IHandler } from '../';

export type CognitoUserPoolHandler = IHandler<
  CognitoUserPoolTriggerEvent,
  CognitoUserPoolTriggerEvent
>;
