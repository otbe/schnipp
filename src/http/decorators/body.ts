import { createMethodParameterDecorator } from '../factories/createMethodParamDecorator';

export const Body = () =>
  createMethodParameterDecorator(e => JSON.parse(e.body || '{}'));
