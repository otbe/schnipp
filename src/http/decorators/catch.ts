import { inject } from 'simple-ts-di';

const CATCH = Symbol('catch');

export const Catch = (...exceptions: any[]): ClassDecorator => (
  target: any
) => {
  inject()(target);
  Reflect.defineMetadata(CATCH, exceptions, target.constructor);
};

export const getCaughtExceptions = (catcher: any) =>
  Reflect.getMetadata(CATCH, catcher.constructor) || [];
