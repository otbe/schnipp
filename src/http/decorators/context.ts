import { createMethodParameterDecorator } from '../factories/createMethodParamDecorator';

export const Context = () =>
  createMethodParameterDecorator((e, m, context) => context);
