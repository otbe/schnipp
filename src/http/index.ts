import {
  APIGatewayEvent,
  APIGatewayProxyResult,
  CustomAuthorizerEvent,
  CustomAuthorizerResult
} from 'aws-lambda';
import { IHandler } from '..';

export type APIGatewayHandler = IHandler<
  APIGatewayEvent,
  APIGatewayProxyResult
>;

export type AuthorizeHandler = IHandler<
  CustomAuthorizerEvent,
  CustomAuthorizerResult
>;

export { AuthPolicy, HttpVerb, Effect } from './utils/AuthPolicy';
export { createResponse } from './utils/createResponse';
export { MetaData } from './utils/MetaData';
export {
  DecoratedExceptionFilter,
  ExceptionFilter
} from './utils/ExceptionFilter';
export { Guard } from './utils/Guard';
export { DefaultExecutionContext } from './utils/ExecutionContext';
export { Catch } from './decorators/catch';
