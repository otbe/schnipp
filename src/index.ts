import { Context, Handler } from 'aws-lambda';
import { ContainedType, getFromContainer } from './container';

export type IHandler<E, R = void> = {
  handle(event: E, context: Context): Promise<R> | R;
};

let setupPromise: Promise<any> | null = null;
export const createHandler = <E, R>(
  handler: ContainedType<IHandler<E, R>>,
  setupFn?: () => Promise<any>
): Handler<E, R> => async (event: E, context: Context) => {
  context.callbackWaitsForEmptyEventLoop = false;

  if (setupFn) {
    if (setupPromise == null) {
      setupPromise = setupFn();
    }

    await setupPromise;
  }

  const handlerInstance = getFromContainer(handler);
  return await handlerInstance.handle(event, context);
};

export { useContainer } from './container';
