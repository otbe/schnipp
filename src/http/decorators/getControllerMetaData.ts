import { MetaData, Guard, DecoratedExceptionFilter } from '..';
import { Newable } from 'simple-ts-di';
import { RestController } from '../rest';
import { GraphQLController } from '../graphql';
import { ResolverMeta } from '../graphql/decorators/method';

export const CONTROLLER = Symbol('controller');

export type ControllerBaseData = {
  headers: { [key: string]: string };
  filters: Array<DecoratedExceptionFilter>;
  guards: Array<Newable<Guard>>;
  metaData: MetaData;
};

export type GraphQLControllerData = ControllerBaseData & {
  resolvers?: Array<ResolverMeta>;
};

export type RestControllerData = ControllerBaseData & {
  basePath?: string;
};

export type ControllerData = GraphQLControllerData | RestControllerData;

export const getControllerMetaData = (
  controller: typeof RestController | typeof GraphQLController
) => {
  let data: ControllerData | null = Reflect.getMetadata(CONTROLLER, controller);

  if (data == null) {
    data = {
      filters: [],
      guards: [],
      metaData: {},
      headers: {}
    };

    Reflect.defineMetadata(CONTROLLER, data, controller);
  }

  return data;
};
