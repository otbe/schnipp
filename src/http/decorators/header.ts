import { getControllerMetaData } from './getControllerMetaData';
import { getMethodMetaData } from '../factories/createHttpMethodDecorator';

export const Header = (name: string, value: string) => (
  ...args: Array<any>
) => {
  const target = args[0];

  switch (args.length) {
    // class
    case 1:
      const controllerMetaData = getControllerMetaData(target);
      controllerMetaData.headers[name] = value;
      return;
    // method
    case 2:
    case 3:
      const methodMetaData = getMethodMetaData(args[1], target);
      methodMetaData.headers[name] = value;
      return args[2];
    default:
      throw 'not supported target for @Header';
  }
};
