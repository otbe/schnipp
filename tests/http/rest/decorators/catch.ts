import 'reflect-metadata';

import { Module, Bind, Container } from 'simple-ts-di';
import { Catch } from '../../../../src/http';
import { getCaughtExceptions } from '../../../../src/http/decorators/catch';

describe('@Catch', () => {
  class Service {}

  @Catch(Error)
  class Test {
    constructor(public service: Service) {}

    catch() {}
  }

  it('should define caught expcetions', () => {
    expect(getCaughtExceptions(Test)).toMatchSnapshot();
  });

  it('should make filter injectable', async () => {
    class MyModule implements Module {
      init(bind: Bind) {
        bind(Test);
        bind(Service);
      }
    }

    const c = new Container(new MyModule());
    const f = await c.get(Test);
    expect(f.service).toBeTruthy();
    expect(f.service).toBeInstanceOf(Service);
  });
});
