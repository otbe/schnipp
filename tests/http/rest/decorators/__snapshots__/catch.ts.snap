// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@Catch should define caught expcetions 1`] = `
Array [
  [Function],
]
`;
